src/NzbDrone.Core/Update/InstallUpdateService.cs

```diff
@@ -164,13 +219,20 @@ public void Execute(ApplicationUpdateCommand message)
 
             if (latestAvailable != null)
             {
-                InstallUpdate(latestAvailable);
+                ExecuteInstallUpdate(message, latestAvailable);
             }
         }
 
         public void Execute(InstallUpdateCommand message)
         {
-            InstallUpdate(message.UpdatePackage);
+            var latestAvailable = _checkUpdateService.AvailableUpdate();
+
+            if (latestAvailable == null || latestAvailable.Hash != message.UpdatePackage.Hash)
+            {
+                throw new ApplicationException("Unknown or invalid update specified");
```

Comment:
Mentioned this on Hangouts. We took the package sent back from the UI, including url and hash, which would allow anyone with access to the UI to upload any application they want and run it by submitting a package of their own.
This check simple verifies if the latest available update is the one the user clicked on and panics otherwise. It shouldn't happen in normal circumstances.